### PR TITLE
Fix Stripe Level 3 data limit for orders with more than 200 different products

### PR DESCRIPTION
**Pull request by elazzabi**
Fixes [#909](909-gh-Automattic/woocommerce-payments-server)

#### Fix Stripe Level 3 data limit

**Context:**
With Stripe's Level 3 data, we are sending metadata about each order to Stripe. This metadata includes a list of product ids bought, description, price, taxes... With a new limitation by Stripe, the maximum basket size is 200 products. This doesn't impact the majority of orders, but we are fixing it to be compliant with Stripe's API.

Stripe offered two solutions: Split charges into separate payments or aggregate items together. We are going with the second option given it's 1) doesn't take much effort (compared to the other one), and 2) this doesn't touch 99% of cases, and thus not worth the effort.

**Solution:**
This change bundles all products (when basket size is > 200) into one product. I _personally think_ this is enough, but I can re-work this PR if it's not the case.

The old behavior is kept intact, which means we'll only change the logic if the basket size is more than 200 different products. This change contains necessary tests too.

#### Testing instructions

* This is unit tested. No manual tests are needed (unless we can have an order with 200 different products)

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) : _Add link here / 'QA Testing Not Applicable'_

### CODE REVIEW

File: includes/class-wc-payment-gateway-wcpay.php [hunk 1]
```diff
@@ -2768,4 +2773,33 @@ public function get_connection_url() {
...
+				function( $cost, $qty ) {
+					return $cost * $qty;
+				},
+				array_column( $items, 'unit_cost' ),
+				array_column( $items, 'quantity' )
+			)
+		);
+
+		return (object) [
+			'product_code'        => (string) substr( uniqid(), 0, 26 ),
```
Comment (elazzabi) [inline includes/class-wc-payment-gateway-wcpay.php:2797]: Random `product_code` given this "bundle" is created programmatically and not tied to a specific bundle

**Review by anu-rock — COMMENTED** [review]
The code changes look great, thanks 👏

But I wonder if we also considered the approach of reducing/bundling only the "excess" items—items beyond 199—into a single item so that the total count would be 200. One benefit is that metadata for the first 199 items will at least not be lost for such big orders.

I'll concede that I'm not sure how often we have to process orders that huge. So the approach you have here might work just fine if it's a rare occurrence.

File: includes/class-wc-payment-gateway-wcpay.php [hunk 2]
```diff
@@ -2128,6 +2128,11 @@ public function get_level3_data_from_order( WC_Order $order ): array {
 		};
 		$items_to_send = array_map( $process_item, $order_items );
 
+		if ( count( $items_to_send ) > 200 ) {
+			// Stripe API max items is 200. If more than that is present, bundle everything in one "product bundle".
+			$items_to_send = [ $this->bundle_level3_data_from_items( $items_to_send ) ];
```
Comment (kalessil) [inline includes/class-wc-payment-gateway-wcpay.php:2133]: I'd suggest keeping the first 199 items unchanged and grouping the rest with your approach (hiding behind 'X items more'). how does it sound?

**Review by kalessil — CHANGES_REQUESTED** [review]

File: includes/class-wc-payment-gateway-wcpay.php (hunk 2 above)
Comment (elazzabi) [reply includes/class-wc-payment-gateway-wcpay.php:2133]: Same as @anu-rock 's suggestion. I think that makes sense!

I'll work on it 👍

File: tests/unit/test-class-wc-payment-gateway-wcpay.php [hunk 3]
```diff
@@ -832,6 +832,9 @@ public function test_level3_data_bundle() {
 		];
 
 		$bundle_data = $this->wcpay_gateway->bundle_level3_data_from_items( $items );
+
+		$this->assertEquals( $bundle_data->product_description, '2 more items' );
```
Comment (anu-rock) [inline tests/unit/test-class-wc-payment-gateway-wcpay.php:836]: We prefer `assertSame` to account for type :)

**Review by anu-rock — APPROVED** [review]
LGTM with the new approach 🎉

File: tests/unit/test-class-wc-payment-gateway-wcpay.php (hunk 3 above)
Comment (kalessil) [reply tests/unit/test-class-wc-payment-gateway-wcpay.php:836]: +1 for using `assertSame`

**Review by kalessil — APPROVED** [review]
LGTM!

File: tests/unit/test-class-wc-payment-gateway-wcpay.php (hunk 3 above)
Comment (elazzabi) [reply tests/unit/test-class-wc-payment-gateway-wcpay.php]: Sounds good! Added in https://github.com/Automattic/woocommerce-payments/pull/3789/commits/5e88b591f6e296ea651c495b23fb219fa1e08592

**Review by dpaun1985 — APPROVED** [review]
LGTM

**Pull request merged**